Add convenience function mapFrom.pick

codeandcats opened the pull request:
Adds a `pick` convenience method to the `mapFrom` utility, which returns a schema that simply picks the given property keys as is. This just DRYs up the mapping of many pass-through properties.

For example, this:

```ts
ObjectMapper.Create<Input, Output>({
  id: 'id',
  name: 'name',
  aliveAt: 'aliveAt',
  brightness: 'brightness',
  externalId: 'externalId',
  geoLocation: 'geoLocation',
  timezone: 'timezone',
  variant: 'variant',
  volume: 'volume',
  hardwareAccelerationEnabled: 'hardwareAccelerationEnabled',
  minimized: 'minimized',
  locked: 'locked',
  isTest: 'isTest',
  targetReleaseAfter: 'targetReleaseAfter',
  targetReleaseTimezone: 'targetReleaseTimezone',
  programmaticPackageType: 'programmaticPackageType',
  createdAt: 'createdAt',
  updatedAt: 'updatedAt',
  deletedAt: 'deletedAt',
  ...other mapping...
})
```

Would becomes this:

```ts
ObjectMapper.Create<Input, Output>({
  ...mapFrom.pick(
    'id',
    'name',
    'aliveAt',
    'brightness',
    'externalId',
    'geoLocation',
    'timezone',
    'variant',
    'volume',
    'hardwareAccelerationEnabled',
    'minimized',
    'locked',
    'isTest',
    'targetReleaseAfter',
    'targetReleaseTimezone',
    'programmaticPackageType',
    'createdAt',
    'updatedAt',
    'deletedAt'
  ),
  ...other mapping...
})
```

When I proposed this yesterday I believe you said the `mapFrom` utility was more designed for use with an individual property. I don't think that needs to be the case though, and I don't think it would be confusing if it wasn't, but that's just MHO. If you feel strongly about it, I can move it somewhere else.

Note I also had a short crack at picking nested paths (e.g. `'name', 'age', ['address', 'country']`) but wasn't able to get that working in the short time I had free today so left that for now but may revisit at a future date.

<details>
  <summary>Side note</summary>

  I actually think that some of the other methods on `mapFrom` could also *optionally* accept multiple keys in order to return a schema with those keys. For example:
  
  ```ts
  ObjectMapper.create<Input, Output>({
    ...mapFrom.omit('foo', 'bar'),
    ...mapFrom.pick('baz', 'qux', 'quux'),
    ...mapFrom.null('corge', 'grault'),
  })
  ```
  
  Would be the same as:
  
  ```ts
  ObjectMapper.create<Input, Output>({
    foo: mapFrom.omit(),
    bar: mapFrom.omit(),
    baz: 'baz',
    qux: 'qux',
    quux: 'quux',
    corge: mapFrom.null(),
    grault: mapFrom.null(),
  })
  ```
  
  Just a thought, which I don't want to distract from what this PR is trying to achieve.
</details>

codeandcats committed:
Add convenience function mapFrom.pick

laurence-myers commented on src/map-from-async.ts:
````diff
@@ -62,4 +63,17 @@ export const mapFromAsync = {
+   * }
+   * ```
+   */
+  pick<TKey extends PropertyKey>(...keys: TKey[]) {
````
🙏 Since this doesn't do anything different from the sync version, I think we can leave this out of `mapFromAsync`. Users can just import it from `mapFrom`.

codeandcats replied:
☑️ Removed `pick` method from `mapFromAsync` 👌

laurence-myers commented on src/map-from.ts:
````diff
@@ -62,4 +62,20 @@ export const mapFrom = {
+   * }
+   * ```
+   */
+  pick<TKey extends PropertyKey>(...keys: TKey[]) {
````
🐛 `mapFrom.pick()` can be called with zero arguments. _(There are "stupid type tricks" you can do to ensure at least one arg is provided)_
🙏 Can we change `keys` to be a single array argument, rather than variadiac args? _(This would require similar but different type tricks, to require at least one array element)_

Also:
⛏️ Add the return type to the function signature.
⛏️ Keep the `mapFrom` items in alphabetical order (`mapFrom.pick` should be before `mapFrom.undefined`)

codeandcats replied:
☑️ Used "stupid type trick" 😄
☑️ Added explicit return type
☑️ Re-ordered methods alphabetically

laurence-myers commented on test/object-mapper.test.ts:
```diff
@@ -377,6 +377,53 @@ describe(ObjectMapper.name, () => {
   });
 
+  it(`allows picking properties`, () => {
+    interface Input {
```
🙏 Could you please add some more "negative" test scenarios to demonstrate that type safety still applies to the picked items? E.g. `firstName: string | undefined` can't be picked into `firstName: string`. (Doesn't need to be comprehensive, just check that the input property's type must be equivalent to, or a subset of, the output property's type.)

codeandcats replied:
☑️ Added negative test scenarios

laurence-myers commented on test/async-object-mapper.test.ts:
```diff
@@ -385,6 +385,53 @@ describe(AsyncObjectMapper.name, () => {
+      firstName: 'Shonky',
+      middleNames: ['The', 'Artful'],
+      lastName: 'Dodger',
+      age: 30,
```
![oh-stop-it-you](https://github.com/user-attachments/assets/61f47a06-a9c5-4d10-bec7-0c28ba6142a8)

codeandcats replied:
<img src="https://media.tenor.com/CqImYid_v7wAAAAM/lie-lee-mack.gif" width="500" /> 
😄

laurence-myers reviewed (changes requested):
👍 Thanks for adding this! Looks good, it'll be very useful. 😊

I've requested some small bits of tidy-up.

🙏 Please run `deno fmt` to fix up the formatting inconsistencies (string characters, additional line breaks)

codeandcats commented:
@laurence-myers I believe I've addressed all your feedback, but if I missed any or you notice something else please let me know and I'll fix it up. I have some time over the break.

Also, if there are other tasks you'd like assistance with, you may like to create a project / issues for and slap on a label like `help-wanted`. I'd be happy to help out. 😃

codeandcats commented:
*Gentle poke* 😄 

<img src="https://gifdb.com/images/high/mr-burns-tickling-poke-pillsbury-doughboy-i9hvjekoz3z3fo0r.gif" width="400" />

laurence-myers reviewed (approved):
<img width="498" height="498" alt="Image" src="https://github.com/user-attachments/assets/80a45cd1-24e7-4eee-a9c9-f494b28e7c82" />

Thanks for adding this. 🙂 Sorry it took so long. 😓

(I got discouraged because I couldn't work out how to solve #2 . Thankfully, Claude has provided a solution, so I'm no longer blocked. 🎉)
